scantool/diag_l2_kwp6227.c

```diff
@@ -0,0 +1,286 @@
+		return rmsg;
+	}
+	if (rmsg == NULL)
+		*errval = DIAG_ERR_NOMEM;
+
+	return rmsg;
+}
+
+static int
+dl2p_6227_startcomms(struct diag_l2_conn *d_l2_conn, flag_type flags,
+	unsigned int bitrate, target_type target, source_type source)
+{
+	struct diag_serial_settings set;
+	struct diag_l2_kwp6227 *dp;
+	int rv;
+	struct diag_l1_initbus_args in;
+
+	if (!(d_l2_conn->diag_link->l1flags & DIAG_L1_DOESFULLINIT) || !(d_l2_conn->diag_link->l1flags & DIAG_L1_DOESL2CKSUM)) {
+		fprintf(stderr, "Can't do KWP6227 on this L0 interface yet, sorry.\n");
+		return diag_iseterr(DIAG_ERR_PROTO_NOTSUPP);
+	}
+
+	if ((flags & DIAG_L2_TYPE_INITMASK) != DIAG_L2_TYPE_SLOWINIT)
+		return diag_iseterr(DIAG_ERR_INIT_NOTSUPP);
+
+	if (diag_calloc(&dp, 1))
+		return diag_iseterr(DIAG_ERR_NOMEM);
+
+	d_l2_conn->diag_l2_proto_data = (void *)dp;
+
+	if (source != 0x13)
+		fprintf(stderr, "Warning : Using tester address %02X. Some ECUs require tester address to be 13.\n", source);
+
+	dp->srcaddr = source;
+	dp->dstaddr = target;
+
+	if (bitrate == 0)
+		bitrate = 10400;
+	d_l2_conn->diag_l2_speed = bitrate;
+
+	set.speed = bitrate;
+	set.databits = diag_databits_8;
+	set.stopbits = diag_stopbits_1;
+	set.parflag = diag_par_n;
+
+	if ((rv=diag_l2_ioctl(d_l2_conn, DIAG_IOCTL_SETSPEED, (void *) &set)))
+		goto err;
+
+	(void)diag_l2_ioctl(d_l2_conn, DIAG_IOCTL_IFLUSH, NULL);
+	diag_os_millisleep(300);
+
+	in.type = DIAG_L1_INITBUS_5BAUD;
+	in.addr = with_parity(target, PARITY_ODD);
+	in.testerid = dp->srcaddr;
+	in.kb1 = 0; in.kb2 = 0;
+	rv = diag_l2_ioctl(d_l2_conn, DIAG_IOCTL_INITBUS, &in);
+	if (rv < 0)
+		goto err;
+
+	if (in.kb1 == 0 && in.kb2 == 0) {
+		d_l2_conn->diag_l2_kb1 = 0xd3;
+		d_l2_conn->diag_l2_kb2 = 0xb0;
+		fprintf(stderr, FLFMT "_startcomms : L0 didn't return keybytes, continuing anyway\n", FL);
+	} else {
+		d_l2_conn->diag_l2_kb1 = in.kb1;
+		d_l2_conn->diag_l2_kb2 = in.kb2;
+	}
+
+	if ((d_l2_conn->diag_l2_kb1 != 0xd3) || (d_l2_conn->diag_l2_kb2 != 0xb0)) {
+		fprintf(stderr, FLFMT "_startcomms : wrong keybytes %02X%02X, expecting D3B0\n",
+			FL, d_l2_conn->diag_l2_kb1, d_l2_conn->diag_l2_kb2);
+		rv = DIAG_ERR_WRONGKB;
+		goto err;
+	}
+
+	return 0;
+
+err:
+	free(dp);
+	d_l2_conn->diag_l2_proto_data = NULL;
+	return diag_iseterr(rv);
+}
+
+static int
+dl2p_6227_stopcomms(struct diag_l2_conn* pX)
+{
+	struct diag_msg msg = {0};
+	uint8_t data[] = { 0xa0 };
+	int errval = 0;
+	static struct diag_msg *rxmsg;
+
+	msg.len = 1;
+	msg.dest = 0; msg.src = 0;	/* use default addresses */
+	msg.data = data;
+
+	rxmsg = dl2p_6227_request(pX, &msg, &errval);
+
+	if (rxmsg == NULL || errval) {
+		fprintf(stderr, "StopDiagnosticSession request failed, waiting for session to time out.\n");
+		diag_os_millisleep(5000);
+	}
+
+	if (rxmsg != NULL)
+		diag_freemsg(rxmsg);
+
+	if (pX->diag_l2_proto_data) {
+		free(pX->diag_l2_proto_data);
+		pX->diag_l2_proto_data=NULL;
+	}
+
+	return 0;
+}
+
+static void
+dl2p_6227_timeout(struct diag_l2_conn *d_l2_conn)
+{
```

Comment:
In the current conditions, this is never called (L0_elm sets the L1_DOESKEEPALIVE flag) ! But it can stay there , in case this L2 driver is modified to also accept L0_dumb interfaces.